cli/crates/typed-resolvers/src/analyze.rs

```diff
@@ -11,209 +13,247 @@ pub(crate) fn analyze<'doc>(graphql_document: &'doc ast::Document<'doc, &'doc st
         }
     }
 
     schema.object_fields.sort_by_key(|(object_id, _)| *object_id);
+    schema.object_field_args.sort_by_key(|(object_id, _)| *object_id);
+}
+
+#[derive(Debug)]
+pub(crate) struct GraphqlType {
+    pub(crate) kind: TypeKind,
+    inner_is_nullable: bool,
+    // TODO: a more compact and/or normalized representation
+    /// The list wrapper types, from innermost to outermost.
+    list_wrappers: Vec<ListWrapper>,
+}
+
+impl GraphqlType {
+    fn resolve(ty: &ast::Type, schema: &AnalyzedSchema<'_>) -> Option<Self> {
+        let (name, inner_is_nullable, list_wrappers) = type_from_nested(ty);
+        Some(GraphqlType {
+            inner_is_nullable,
+            list_wrappers,
+            kind: BuiltinScalar::from_str(name)
+                .ok()
+                .map(TypeKind::BuiltinScalar)
+                .or_else(|| schema.definition_names.get(name).map(|d| TypeKind::Definition(*d)))?,
+        })
+    }
+
+    /// Returns whether the innermost type is nullable.
+    ///
+    /// For example `[Test]!` would return true, `[Test!]` would return false, `Test` would be
+    /// true, and `Test!` would be false.
+    pub(crate) fn inner_is_nullable(&self) -> bool {
+        self.inner_is_nullable
+    }
+
+    /// Iterate list wrapper types from innermost to outermost.
+    ///
+    /// Example: `[[Int!]!]` would yield `NonNull, NonNullList, List`. You can then look up the
+    /// unwrapped type in `OutputField::kind`.
+    pub(crate) fn iter_list_wrappers(&self) -> impl Iterator<Item = ListWrapper> + '_ {
+        self.list_wrappers.iter().copied()
+    }
 }
 
 fn analyze_ast_input_field<'doc>(
-    field: &'doc ast::InputValue<'doc, &'doc str>,
+    field: &'doc ast::InputValueDefinition,
     schema: &AnalyzedSchema<'doc>,
 ) -> Option<Field<'doc>> {
-    let (name, inner_is_nullable, list_wrappers) = type_from_nested(&field.value_type);
     Some(Field {
-        name: field.name,
-        docs: field.description.as_deref(),
-        kind: BuiltinScalar::from_str(name)
-            .ok()
-            .map(FieldTypeKind::BuiltinScalar)
-            .or_else(|| schema.definition_names.get(name).map(|d| FieldTypeKind::Definition(*d)))?,
-        inner_is_nullable,
-        list_wrappers,
+        name: field.name.node.as_str(),
+        docs: field.description.as_ref().map(|d| d.node.as_str()),
+        r#type: GraphqlType::resolve(&field.ty.node, schema)?,
+        resolver_name: None, // no resolvers on input fields
     })
 }
 
-fn analyze_ast_field<'doc>(
-    field: &'doc ast::Field<'doc, &'doc str>,
-    schema: &AnalyzedSchema<'doc>,
-) -> Option<Field<'doc>> {
-    let (name, inner_is_nullable, list_wrappers) = type_from_nested(&field.field_type);
+fn analyze_ast_field<'doc>(field: &'doc ast::FieldDefinition, schema: &AnalyzedSchema<'doc>) -> Option<Field<'doc>> {
+    let resolver_name = field
+        .directives
+        .iter()
+        .find(|directive| directive.node.name.node == "resolver")
```

Comment:
This stuff from the graphql parser is super annoying. We don't need it here, but for the schema registry we should hide this `directive.node.name.node` dance...

Reply:
Yes we won't want to work with ASTs directly, we'll have something like `ParserDatabase` for things like checks.